dev/reference/performance/optimizer-hints.md

```diff
@@ -9,51 +9,227 @@ TiDB 支持 Optimizer Hints 语法，它基于 MySQL 5.7 中介绍的类似 comm
 
-### TIDB_INLJ(t1, t2)
+## 优化器相关 Hint 语法
+
+优化器相关的 Hint 可以以注释的形式写在**任意**一个`SELECT`后面。Hint 的生效范围以及指定表的生效范围可通过下面要介绍的 Query Block 来指定。若不指定 Query Block，默认生效范围为当前所在`SELECT`的 Query Block 内。
+
+### Query Block
+
+一条语句中每一个查询和子查询都对应着一个不同的 Query Block，每个 Query Block 有自己对应的 QB_NAME。
+以下面这条语句为例：
+
+{{< copyable "sql" >}}
+
+```sql
+select * from (SELECT * from t) t1, (SELECT * from t) t2;
+```
+
+该查询语句有 3 个 Query Block，最外面一层`SELECT`所在的 Query Block 的 QB_NAME 为`sel_1`，里面两个子查询的 QB_NAME 依次为`sel_2`和`sel_3`。
+
+### QB_NAME
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ QB_NAME(QB1) */ * from t;
+```
+
+将当前 Query Block 的 QB_NAME 设为指定值，同时原本的 QB_NAME（在该例子中是 sel_1）仍然有效的。但是如果指定的 QB_NAME 为 sel_2，并且不给原本的 sel_2 指定新的 QB_NAME，将不再能正确地指向原本的 sel_2。
+
+### 参数
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ HASH_JOIN(@sel_1 t1@sel_1, t2) */ * from t t1, t t2 where t1.a = t2.a;
+```
+
+优化器相关的 Hint 中除 `QB_NAME` 外都可以通过可选参数 @QB_NAME 来指定生效范围。该参数需写在最前面，与其他参数用空格隔开。
+
+参数中的每一个表名都可以在后面加 @QB_NAME 来指定是哪个 Query Block 中的表。
+
+### SM_JOIN(t1, t2)
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ SM_JOIN(t1, t2) */ * from t1，t2 where t1.id = t2.id;
+```
+
+提示优化器使用 Sort Merge Join 算法。这个算法通常会占用更少的内存，但执行时间会更久。当数据量太大，或系统内存不足时，建议尝试使用。
+
+别名：TIDB_SMJ (3.0 及以下版本仅支持使用该别名)
+
+### INL_JOIN(t1, t2)
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ INL_JOIN(t1, t2) */ * from t1，t2 where t1.id = t2.id;
+```
+
+提示优化器使用 Index Nested Loop Join 算法。这个算法可能会在某些场景更快，消耗更少系统资源，有的场景会更慢，消耗更多系统资源。对于外表经过 WHERE 条件过滤后结果集较小（小于 1 万行）的场景，可以尝试使用。`TIDB_INLJ()` 中的参数是建立查询计划时，内表的候选表。即 `TIDB_INLJ(t1)` 只会考虑使用 t1 作为内表构建查询计划。
+
+别名：TIDB_INLJ (3.0 及以下版本仅支持使用该别名)
+
+### HASH_JOIN(t1, t2)
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ HASH_JOIN(t1, t2) */ * from t1，t2 where t1.id = t2.id;
+```
+
+提示优化器使用 Hash Join 算法。这个算法多线程并发执行，执行速度较快，但会消耗较多内存。
+
+别名：TIDB_HJ (3.0 及以下版本仅支持使用该别名)
+
+### HASH_AGG()
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ HASH_AGG() */ count(*) from t1，t2 where t1.a > 10 group by t1.id;
+```
+
+提示优化器使用 Hash Aggregation 算法。这个算法多线程并发执行，执行速度较快，但会消耗较多内存。
+
+### STREAM_AGG()
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ STREAM_AGG() */ count(*) from t1，t2 where t1.a > 10 group by t1.id;
+```
+
+提示优化器使用 Stream Aggregation 算法。这个算法通常会占用更少的内存，但执行时间会更久。数据量太大，或系统内存不足时，建议尝试使用。
+
+### USE_INDEX(t1, idx1, idx2)
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ USE_INDEX(t1, idx1, idx2) */ * from t t1;
+```
+
+提示优化器对指定表仅使用给出的索引。效果等价于 `select * from t t1 index(idx1, idx2);`。
+
+### IGNORE_INDEX(t1, idx1, idx2)
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ USE_INDEX(t1, idx1, idx2) */ * from t t1;
+```
+
+提示优化器对指定表忽略给出的索引。效果等价于 `select * from t t1 ignore index(idx1, idx2);`。
+
+### AGG_TO_COP()
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ AGG_TO_COP() */ sum(t1.a) from t t1;
+```
+
+提示优化器将聚合操作下推到 coprocessor。如果优化器没有下推某些适合下推的聚合函数，建议尝试使用。
+
+### READ_FROM_STORAGE()
+
+{{< copyable "sql" >}}
+
+```sql
+select /*+ READ_FROM_STORAGE() */ t1.a from t t1;
```

Comment:
Point out the accepted storages?

Reply:
What is the behavior when param is empty?

Reply:
when parameter is empty, it will trigger a warning because of parser syntax error.